### PR TITLE
Add support for mock function type overrides.

### PR DESCRIPTION
**Pull request by tachyonics**
*Issue #, if available:*

*Description of changes:*

Add support for mock function type overrides.

TLDR: The upshot of these changes is an application should be able to add an sync override to the mock or throwing implementations with Swift 5.4 and it will be treated as an async function by the compiler in Swift 5.5 without any application code changes (and could be converted to an async function sometime later if required).

If code generated with the asyncAwaitGeneration:client set to NONE (currently hard-coded by some-framework-application-generate)-
1. Will create an alternate set of override input functions for the mock and mock throwing client implementations that are sync functions
2. The mock and throwing implementations call into the functions added here [1] and [2]. These will use the new override functions if they exist, otherwise use the existing EventLoopFuture-returning overrides if they exist, otherwise return a default or throw an error depending on the implementation

This is an example of the code generated changes - [3]

If code generated with the asyncAwaitGeneration:client set to EXPERIMENTAL-
1. Will create an alternate set of override input functions for the mock and mock throwing client implementations that are sync functions if compiled under Swift 5.4 or earlier or async functions of compiled under Swift 5.5 or later (and supporting Apple platforms)
2. The mock and throwing implementations call into the functions added here [1] and [2] for Swift 5.4 and earlier and [4] and [5] for Swift 5.5 and later. These will use the new override functions if they exist, otherwise use the existing EventLoopFuture-returning overrides if they exist, otherwise return a default or throw an error depending on the implementation.

This is an example of the code generated changes - [6]

*Testing:*

Tested application consuming a client and using a sync function override both with asyncAwaitGeneration:client set to NONE and EXPERIMENTAL.
1. Both variants of the client compiled under 5.4.1 and 5.5 nightly without any application changes.
2. The asyncAwaitGeneration:client set to EXPERIMENTAL client compiled under Swift 5.5 allowed the function override to be made async by adding the `async` keyword.

[1] https://github.com/amzn/smoke-aws/blob/mock_and_mock_throwing_implementations/Sources/SmokeAWSHttp/MockClientProtocol.swift
[2] https://github.com/amzn/smoke-aws/blob/mock_and_mock_throwing_implementations/Sources/SmokeAWSHttp/MockThrowingClientProtocol.swift
[3] https://github.com/amzn/smoke-framework-examples/commit/bb77a4728606551c6de2aa6ae8e3229690316fa8
[4] https://github.com/amzn/smoke-aws/blob/mock_and_mock_throwing_implementations/Sources/_SmokeAWSHttpConcurrency/MockClientProtocol.swift
[5] https://github.com/amzn/smoke-aws/blob/mock_and_mock_throwing_implementations/Sources/_SmokeAWSHttpConcurrency/MockThrowingClientProtocol.swift
[6] https://github.com/amzn/smoke-framework-examples/commit/c5c50f65a721260e0e44a30e3e23e9ab44492044

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: Sources/ServiceModelGenerate/MockClientDelegate.swift [hunk 1]
```diff
@@ -45,17 +45,20 @@ public struct MockClientDelegate: ModelClientDelegate {
         self.asyncAwaitGeneration = asyncAwaitGeneration
         
         let name: String
+        let additionalConformingProtocol: String
```
Comment (Josh-Gearou) [inline Sources/ServiceModelGenerate/MockClientDelegate.swift:48]: I suppose if we added a third protocol we would rename this string, as "additional" wouldn't really make sense then?  Might be better to rename this now, but will leave this up to you.

**Review by Josh-Gearou — APPROVED** [review]

**Review by Josh-Gearou — DISMISSED** [review]

**Review by Josh-Gearou — DISMISSED** [review]

File: Sources/ServiceModelGenerate/MockClientDelegate.swift (hunk 1 above)
Comment (tachyonics) [reply Sources/ServiceModelGenerate/MockClientDelegate.swift]: Yup, renamed.

**Review by Josh-Gearou — APPROVED** [review]
